openwisp_controller/config/api/serializers.py

```diff
@@ -0,0 +1,205 @@
+import ast
+import collections
+import json
+
+from rest_framework import serializers
+from swapper import load_model
+
+from openwisp_users.api.mixins import FilterSerializerByOrgOwned
+from openwisp_utils.api.serializers import ValidatedModelSerializer
+
+Template = load_model('config', 'Template')
+Vpn = load_model('config', 'Vpn')
+Device = load_model('config', 'Device')
+Config = load_model('config', 'Config')
+TemplateTag = load_model('config', 'TemplateTag')
+Organization = load_model('openwisp_users', 'Organization')
+
+
+class BaseMeta:
+    read_only_fields = ['created', 'modified']
+
+
+class TemplateSerializer(FilterSerializerByOrgOwned, ValidatedModelSerializer):
+    config = serializers.JSONField()
+    tags = serializers.StringRelatedField(many=True, read_only=True)
+    default_values = serializers.JSONField(required=False, default={})
+
+    class Meta(BaseMeta):
+        model = Template
+        fields = [
+            'id',
+            'name',
+            'tags',
+            'organization',
+            'type',
+            'backend',
+            'default',
+            'required',
+            'default_values',
+            'config',
+            'created',
+            'modified',
+        ]
+
+
+class VpnSerializer(FilterSerializerByOrgOwned, ValidatedModelSerializer):
+    config = serializers.JSONField()
+
+    class Meta(BaseMeta):
+        model = Vpn
+        fields = [
+            'id',
+            'name',
+            'host',
+            'organization',
+            'key',
+            'ca',
+            'cert',
+            'backend',
+            'notes',
+            'dh',
+            'config',
+        ]
+
+
+class JsonConfigField(serializers.Field):
+    def to_representation(self, value):
+        return json.loads(json.dumps(value))
+
+    def to_internal_value(self, data):
+        if type(data) is str:
+            return ast.literal_eval(data)
+        else:
+            return data
+
+
+class JsonContextField(serializers.Field):
+    def to_representation(self, value):
+        return dict(value)
+
+    def to_internal_value(self, data):
+        if type(data) is str:
+            return ast.literal_eval(data)
+        else:
+            return data
+
+
+class DeviceConfigSerializer(serializers.ModelSerializer):
+    config = JsonConfigField(
+        help_text='''<i>config</i> field in HTML form
+        accepts values in dictionary format'''
+    )
+    context = JsonContextField(
+        help_text='''<i>context</i> field in HTML form
+        accepts values in dictionary format'''
+    )
+
+    class Meta:
+        model = Config
+        fields = ['backend', 'status', 'templates', 'context', 'config']
+        extra_kwargs = {'status': {'read_only': True}}
+
+
+class DeviceListSerializer(serializers.ModelSerializer):
```

Comment:
We can't use `ValidatedModelSerializer` here, because since this validation is designed to validate a single level Model serializer but this is a Nested serializer, here when I am using it after the validation is done it returns only the data of the parent serializer opting out the data of nestedserializer, resulting in an exception error.